pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -1648,6 +1740,84 @@ func (gce *GCECloud) ListClusters() ([]string, error) {
 	return result, nil
 }
 
+func (gce *GCECloud) ListClusters() ([]string, error) {
+	allClusters := []string{}
+
+	for _, zone := range gce.managedZones {
+		clusters, err := gce.listClustersInZone(zone)
+		if err != nil {
+			return nil, err
+		}
+		// TODO: Scoping?  Do we need to qualify the cluster name?
+		allClusters = append(allClusters, clusters...)
+	}
+
+	return allClusters, nil
+}
+
 func (gce *GCECloud) Master(clusterName string) (string, error) {
 	return "k8s-" + clusterName + "-master.internal", nil
 }
+
+type gceInstance struct {
+	Zone  string
+	Name  string
+	ID    uint64
+	Disks []*compute.AttachedDisk
+}
+
+type gceDisk struct {
+	Zone string
+	Name string
+	Kind string
+}
+
+func (gce *GCECloud) resolveHosts(hostNames []string) ([]*gceInstance, error) {
+	// TODO: More efficient implementation
+	// I think a good implementation might be:
+	//  * pass multiple instances to each zone (see computeHostTags)
+	//  * check each zone in turn (in parallel?)
+	//  * try local zone first (so we don't pay the tax unless we have to)
+	//  * remove instances from the list once we've found them
+	//  * we could cache 'last known zone' by name, but I don't know if this will be needed
```

Comment:
Yes, I think we need to improve this, as right now it's O(instances x zones) in the number or remote calls it makes (we could have thousands of instances, so that's real bad).

A pretty simple improvement to reduce it to O(zones) would be to switch around resolveHost() to call resolveHosts() rather than the other way around.  The heavy work would get done by resolveHosts (which would look very much like the current resolveHost(), iterating over all zones, getting all matching instances from each of them (with a server-side filter).